### ISSUE TITLE
Uploading Project Avatar

### ISSUE THREAD

**Issue by oliveirarafa** [gitlabform/gitlabform#698]
I'm having trouble uploading avatars for my projects. I always get `{"error":"avatar is invalid"}` error.

I can't change using curl to change one by one.

I think that the problem is that gitlabform can't find my image file but can't figure out the correct url. Tried relative, absolute, with starting '@'...

**Comment by amimas** [comment]
Hi. Could you please share a sample config you have tried?

**Comment by oliveirarafa** [comment]
> Hi. Could you please share a sample config you have tried?

Yeah, no problem.

```yml
  project_settings:
      avatar: './my-photo.jpg',
      description: |
        Projeto de documentação
```

I tried putting a base64 image, svg, png. Searched some forums and the gitlab api docs but couldnt figure out the format that I should use.

**Comment by amimas** [comment]
Thanks for the sample config. What you tried won't work, regardless of value format. It comes down to Gitlab's API.

 The `project_settings` config uses [Project API](https://docs.gitlab.com/ee/api/projects.html#edit-project) as per the [gitlabform documentation](https://gitlabform.github.io/gitlabform/reference/settings/). This API doesnt handle the project avatar.

[Project avatar](https://docs.gitlab.com/ee/api/projects.html#upload-a-project-avatar) is managed by a different gitlab API. Gitlabform currently doesn't have support for it. PR is welcome though.

Would you be interested in implementing it? Let's discuss what the config should be like for this setting. This feature probably should also support deleting a project avatar.

**Comment by oliveirarafa** [comment]
> Thanks for the sample config. What you tried won't work, regardless of value format. It comes down to Gitlab's API.
> 
> The `project_settings` config uses [Project API](https://docs.gitlab.com/ee/api/projects.html#edit-project) as per the [gitlabform documentation](https://gitlabform.github.io/gitlabform/reference/settings/). This API doesnt handle the project avatar.
> 
> [Project avatar](https://docs.gitlab.com/ee/api/projects.html#upload-a-project-avatar) is managed by a different gitlab API. Gitlabform currently doesn't have support for it. PR is welcome though.
> 
> Would you be interested in implementing it? Let's discuss what the config should be like for this setting. This feature probably should also support deleting a project avatar.

Oh, ok. Thank you. Do you know what is the avatar prop in [Project API](https://docs.gitlab.com/ee/api/projects.html#edit-project)? I think their documentation needs a fix, it kind of implies that it is possible to change the avatar from this api.

**Comment by amimas** [comment]
Hmm... I didn't realize earlier that the "Edit project" API endpoint had an `avatar` property. I was curious how that works. So, tried a few things using `curl`. That API endpoint does allow updating project avatar like this:

```sh
curl --silent --request PUT --header "PRIVATE-TOKEN: ${GITLAB_TOKEN}" --url "${GITLAB_API_URL}/projects/<PROJECT_ID>" --form "avatar=@${HOME}/Downloads/sample-logo.jpeg" | jq .
```

The API for "[Upload a project avatar](https://docs.gitlab.com/ee/api/projects.html#upload-a-project-avatar)" works the same way.

So, in theory the config you had is fine. But, I don't think gitlabform has logic to process that config because as you can see the `avatar` requires sending an image file to the API endpoint.

**Comment by oliveirarafa** [comment]
> Hmm... I didn't realize earlier that the "Edit project" API endpoint had an `avatar` property. I was curious how that works. So, tried a few things using `curl`. That API endpoint does allow updating project avatar like this:
> 
> ```shell
> curl --silent --request PUT --header "PRIVATE-TOKEN: ${GITLAB_TOKEN}" --url "${GITLAB_API_URL}/projects/<PROJECT_ID>" --form "avatar=@${HOME}/Downloads/sample-logo.jpeg" | jq .
> ```
> 
> The API for "[Upload a project avatar](https://docs.gitlab.com/ee/api/projects.html#upload-a-project-avatar)" works the same way.
> 
> So, in theory the config you had is fine. But, I don't think gitlabform has logic to process that config because as you can see the `avatar` requires sending an image file to the API endpoint. And right now it's probably just sending the file path as string, which isn't what the api is expecting. That's probably why you're seeing that error about invalid value.

Oh I see. I haven't looked into gitlabform's code but maybe it is possible to use the curl to use the already available edit project endpoint when the avatar urls is used?

I don't know if im have enought knowledge about gitlabform to send a PR. Maybe in the future, but I think it a good feature to have.

Thank you for helping me, @amimas!

**Comment by amimas** [comment]
@rickbrouwer - that's a good point. Looking at gitlab's API doc for removing avatar, apparently you just need to set the attribute to a blank string.

https://docs.gitlab.com/api/projects/?hl=en-CA#remove-a-project-avatar

So it might be as simple as following steps:

* Set project.avatar to '': This is the key step. By setting the avatar attribute of the project object to an empty string, you instruct the GitLab API to remove the current avatar.
 * Save the project: Calling project.save() sends a PUT request to the GitLab API to update the project with the new avatar value (which is empty).

Based on the finding, I wonder if the syntax should be more generic... Maybe a little bit derived from your PR:

```yaml
projects_and_groups:
  group_1/project_1:
    project_settings:
      visibility: internal
      avatar: <path to image file | empty string for deletion>
```

Not sure if the above is intuitive. Any other suggestions?

What do you think? Haven't checked the group setting API. Hopefully there are no surprises. Would you be able to test out the avatar deletion using python-gitlab?

**Comment by amimas** [comment]
@gdubicki - please see the previous comment. The avatar deletion related suggestion goes against what you had suggested in PR #987. I agree with you about consistency with other resources deletion related syntax, but this case seems different based on how Gitlab's API works.

**Comment by rickbrouwer** [comment]
We can also do something like:

```
if isinstance(avatar_config, dict) and avatar_config.get("delete", False):
            debug("Deleting project avatar")
            project.avatar = ""
            project.save()
            debug("Avatar delete successfully")

```

But, i have tested this with an empty string, that does not work.
There is also an issue with this at python-gitlab: https://github.com/python-gitlab/python-gitlab/issues/2933

**Comment by amimas** [comment]
Yeah, we can do that but I'm just thinking about the config syntax. Wondering if it's necessary to have the `delete` key. Going with multiple data type for `avatar` attribute aligns with gitlab's API.

Did you test it using the `project` object? Looking at the issue you mentioned, that seems to be specific to avatar for `topic` object. I didn't know topics could have avatar too - learned something new. We should probably keep that out of scope for now.

Anyways, could you please confirm if removing avatar using `project` works or not?

**Comment by gdubicki** [comment]
> [@gdubicki](https://github.com/gdubicki) - please see the previous comment. The avatar deletion related suggestion goes against what you had suggested in PR [#987](https://github.com/gitlabform/gitlabform/pull/987). I agree with you about consistency with other resources deletion related syntax, but this case seems different based on how Gitlab's API works.

Hey, of course if we switch to use `project_settings`, it’s make sense to set the `avatar` key to an empty string (or null) to delete the avatar. 👍

**Comment by rickbrouwer** [comment]
> Yeah, we can do that but I'm just thinking about the config syntax. Wondering if it's necessary to have the `delete` key. Going with multiple data type for `avatar` attribute aligns with gitlab's API.
> 
> Did you test it using the `project` object? Looking at the issue you mentioned, that seems to be specific to avatar for `topic` object. I didn't know topics could have avatar too - learned something new. We should probably keep that out of scope for now.
> 
> Anyways, could you please confirm if removing avatar using `project` works or not?

Yeah, topics are mentioned there, but the issue also applies to projects and groups.
When I send the avatar as an empty string like:

```
    group_settings:
      avatar:
        file: ""
```

I get the same error as in the issue:

```
Warning: Error occurred while processing group REDACTED/REDACTED/REDACTED, exception:
400: {'avatar': ['file format is not supported. Please try one of the following supported formats: image/png, image/jpeg, image/gif, image/bmp, image/tiff, image/vnd.microsoft.icon']}
```

**Comment by amimas** [comment]
Thanks @rickbrouwer . I would wx

**Comment by rickbrouwer** [comment]
> Any chance you are able to contribute a fix in python-gitlab?

Sure! 😃 https://github.com/python-gitlab/python-gitlab/pull/3175

**Comment by amimas** [comment]
Thank yo so much @rickbrouwer !!
